common/metrics/go-metrics-m3.go

```diff
@@ -0,0 +1,159 @@
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package metrics
+
+import (
+	"fmt"
+	"time"
+
+	gometrics "github.com/rcrowley/go-metrics"
+	"github.com/uber-common/bark"
+)
+
+const (
+	delayToFirstLogDump  = time.Minute * 15
+	delayBetweenLogDumps = time.Minute * 60
+	delayBetweenExports  = time.Second * 10
+)
+
+// goMetricsExporter allows an rcrowley/go-metrics registry to be
+// periodically exported to the Cherami metrics system
+type goMetricsExporter struct {
+	c     Client
+	scope int
+
+	// metricNameToMetric maps a metric name to the metric symbol,
+	// e.g. "request-rate" -> OutputhostCGKafkaRequestRate
+	metricNameToMetric  map[string]int
+	closeCh             <-chan struct{}
+	metricNameToLastVal map[string]int64
+	l                   bark.Logger
+	registry            gometrics.Registry // Registry to be exported
+	lastLogDump         time.Time
+}
+
+// NewGoMetricsExporter starts the exporter loop to export go-metrics to the Cherami
+// metrics system and returns the go-metrics registry that should be used
+func NewGoMetricsExporter(
+	c Client,
+	scope int,
+	metricNameToMetric map[string]int,
+	l bark.Logger,
+	closeCh <-chan struct{},
+) gometrics.Registry {
+	// Make a new registy and give it to the caller. This allows the exporters to operate independently
+	registry := gometrics.NewRegistry()
+	r := &goMetricsExporter{
+		c:                   c,
+		scope:               scope,
+		registry:            registry,
+		metricNameToMetric:  metricNameToMetric,
+		l:                   l,
+		closeCh:             closeCh,
+		metricNameToLastVal: make(map[string]int64),
+
+		// Calculate the time to provide the delayToFirstLogDump in export()
+		lastLogDump: time.Now().Add(-1 * delayBetweenLogDumps).Add(delayToFirstLogDump),
+	}
+	go r.run()
+	return registry
+}
+
+func (r *goMetricsExporter) run() {
+	t := time.NewTicker(delayBetweenExports)
+	defer t.Stop()
+
+exportLoop:
+	for {
+		r.export()
+
+		select {
+		case <-t.C:
+		case <-r.closeCh:
+			break exportLoop
+		}
+	}
+}
+
+func (r *goMetricsExporter) export() {
+	r.registry.Each(func(name string, i interface{}) {
```

Comment:
I am not sure I understand this. So for every metric in the 'go-metrics' registry, we emit into m3? This would mean, we are expected to read in every metric, right?

Wouldn't it be better if probably we only pulled in metrics that we have listed in our metrics map that we pass in with 'New' .. that way we can ignore some metrics that we aren't interested in, etc?